x-pack/elastic-agent/pkg/agent/application/upgrade/error_checker.go

```diff
@@ -0,0 +1,85 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control/client"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
+)
+
+const (
+	statusCheckPeriod = 30 * time.Second
+)
+
+// ErrAgentStatusFailed is returned when agent reports FAILED status.
+var ErrAgentStatusFailed = errors.New("agent in a failed state", errors.TypeApplication)
+
+// ErrorChecker checks agent for status change and sends an error to a channel if found.
+type ErrorChecker struct {
+	notifyChan  chan error
+	log         *logger.Logger
+	agentClient client.Client
+}
+
+// NewErrorChecker creates a new error checker.
+func NewErrorChecker(ch chan error, log *logger.Logger) (*ErrorChecker, error) {
+	c := client.New()
+	ec := &ErrorChecker{
+		notifyChan:  ch,
+		agentClient: c,
+		log:         log,
+	}
+
+	return ec, nil
+}
+
+// Run runs the checking loop.
+func (ch ErrorChecker) Run(ctx context.Context) {
+	ch.log.Debug("Error checker started")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(statusCheckPeriod):
+			err := ch.agentClient.Connect(ctx)
+			if err != nil {
+				ch.log.Error(err, "Failed communicating to running daemon", errors.TypeNetwork, errors.M("socket", control.Address()))
```

Comment:
What happens if the new Agent starts running but there is a bug that never brings up the control socket? This case would not be detected by the PID watcher.

I think the un-ability to communicate with the Agent is something the `ErrorChecker` should report.

Maybe we give more chances for this to fail, as the Agent could be starting, but I think it needs to be reported.

Reply:
Was this solved? Doesn't seem like the failed ability to connect, is handled.